src/Npgsql.EntityFrameworkCore.PostgreSQL/Migrations/Internal/NpgsqlHistoryRepository.cs

```diff
@@ -85,22 +85,33 @@ protected override string ExistsSql
 
         public override string GetCreateIfNotExistsScript()
         {
-            return GetCreateScript();
+            var script = GetCreateScript();
+            return script.Insert(script.IndexOf("CREATE TABLE", StringComparison.Ordinal) + 12, " IF NOT EXISTS");
         }
 
         public override string GetBeginIfNotExistsScript(string migrationId)
         {
-            throw new NotSupportedException("Generating idempotent scripts for migration is not currently supported by Npgsql");
+            string dot = TableSchema != null ? "." : "";
+            return $@"
+DO $$
+BEGIN
+    IF NOT EXISTS(SELECT 1 FROM {TableSchema}{dot}""{TableName}"" WHERE ""{MigrationIdColumnName}"" = '{migrationId}') THEN";
```

Comment:
Almost there...

Rather than including `TableSchema` and `TableName` directly, and having to mess around with the dot, you can just use `SqlGenerationHelper.DelimitIdentifier(TableName, TableSchema)` which does all that for you. You can look at SqlServerHistoryRepository for an example of how they do it.